tactics/hints.ml

```diff
@@ -1189,20 +1208,55 @@ let subst_autohint (subst, obj) =
   in
   if action == obj.hint_action then obj else { obj with hint_action = action }
 
+let is_hint_local = function Local -> true | Export | SuperGlobal -> false
+
 let classify_autohint obj =
-  match obj.hint_action with
-  | AddHints [] -> Dispose
-  | _ -> match obj.hint_local with
-    | Local -> Dispose
-    | Export | SuperGlobal -> Substitute obj
+  if is_hint_local obj.hint_local || is_trivial_action obj.hint_action then Dispose
+  else Substitute obj
+
+let discharge_autohint (_, obj) =
+  if is_hint_local obj.hint_local then None
+  else
+    let action = match obj.hint_action with
+    | AddTransparency { grefs; state } ->
+      let grefs = match grefs with
+      | HintsVariables | HintsConstants -> grefs
+      | HintsReferences grs ->
+        let filter = function
+        | EvalConstRef c -> true
+        | EvalVarRef id -> not @@ Lib.is_in_section (GlobRef.VarRef id)
+        in
+        let grs = List.filter filter grs in
+        HintsReferences grs
+      in
+      AddTransparency { grefs; state }
+    | AddHints _ | RemoveHints _ ->
+      (* not supported yet *)
+      assert false
+    | AddCut path ->
+      if is_section_path path then AddHints [] (* dummy *) else obj.hint_action
```

Comment:
Why dummy instead of None?

Reply:
Pure question of code style, it would have required making some ugly commutations and duplications.